HP-836: optimize graphql queries

NikoHelle opened the pull request:
When users navigate to their profile page, the page loads
- ProfileExistsQuery
- MyProfileQuery
and also
- ServiceConnectionsQuery
- GdprServiceConnectionsQuery 

Neither ServiceConnectionQuery is required when profile is shown. Loading started, because DeleteProfile ja DownloadProfile components used useQuery-hooks instead of useLazyQuery. 

DeleteProfile requires ServiceConnections to be loaded so the UI can show correct information before deleting the profile. Had to add a loader and load indicator. Now the UI loads the connections first and then shows proper warnings. 

DownloadProfile won't show any additional data to user, so lazy query was easy to add there.

NikoHelle committed:
(HP-836) Use only lazy queries in useDownloadProfile

NikoHelle committed:
(HP-836) Change useDeleteProfile lazy
Have to use fetchPolicy: 'no-cache' or onCompleted is never called. This is a ApolloGraphQL bug when data is cached.

NikoHelle committed:
(HP-836) ExpandingPanel can call onChange callback

NikoHelle committed:
(HP-836) Load service connections only when panel is opened

Removed data loading on first component render.
Data is needed only when user opens the panel and will delete profile.

akikoskinen commented on src/profile/components/deleteProfile/DeleteProfile.tsx:
```diff
@@ -52,24 +60,52 @@ function DeleteProfile(): React.ReactElement {
+    }
+  }, [getServiceConnections, setDataLoadState, dataLoadState]);
+
+  const onExpandingPanelChange = useCallback(
+    isOpen => {
+      if (isOpen) {
+        loadServiceConnections();
+      }
+    },
+    [loadServiceConnections]
```
Could the service connections fetching be delayed even more, to the point when the last warning modal is shown?

NikoHelle replied:
Good idea! Fetching **could** be done after user clicks the delete button! 

I tried it... and looks like there is a major accessibility issues there, if the load animation is in the ConfirmationModal. The ConfirmationModal is "accessibility-wise" an aria alert. If the content of the modal is loaded later and changes, it's hard to indicate to screen readers what happened. For example buttons appear after loading. 

Also UX-wise it would seem that pressing the delete button would start the deletion (as the user sees something is loading), but it would only load the ServiceConnections and then ask "Are you sure".  Seems a bit odd and if it takes a long time, user doesn't know was the profile deleted now or not.

Also there are some issues with the Dialog component, if the heading changes while it is open. The heading is auto-focused when Dialog opens and loses focus when heading element changes. Having two different modals would solve this, but is it worth it...?

Too many issues rises, if the fetching is moved, so I'd leave it there. 😄

akikoskinen replied:
Good thoughts. Makes sense to leave the loading where it is now. But could the `t('deleteProfile.explanation')` text be shown already immediately, and the spinner would roll below it? It's quite a long text so the user could already read it while the loading happens.

NikoHelle replied:
Good idea. Done.

akikoskinen commented:
> When users navigate to their profile page, the page loads
> 
>     * ProfileExistsQuery
>     * MyProfileQuery

Not sure if it's part of this PR, but to me it seems that the `ProfileExistsQuery` is needless. It uses the `myProfile` GraphQL query to check if a profile exists. Then, if profile exists, it executes the `MyProfileQuery`, which **again** uses `myProfile` GraphQL query to fetch the profile data. It should be enough to use the `myProfile` GraphQL query only once.

NikoHelle committed:
(HP-836) Prevent 2 simultaneous ApiToken fetches on login

When user logs in, apiToken is fetched. But api token is also fetched when "userLoaded" event happens. For example at login.

From client docs: "addUserLoaded: Raised when a user session has been established (or re-established".

From code: looks like re-establishing means accessToken has changed or user was loaded with getUser.

NikoHelle committed:
(HP-836) Fix user object checks on failed login

When user logs in, but there is an error, the returned user object passed a check on ProfilePage and tried to get profile data and api token.

NikoHelle commented:
> > When users navigate to their profile page, the page loads
> > ```
> > * ProfileExistsQuery
> > * MyProfileQuery
> > ```
> 
> Not sure if it's part of this PR, but to me it seems that the `ProfileExistsQuery` is needless. It uses the `myProfile` GraphQL query to check if a profile exists. Then, if profile exists, it executes the `MyProfileQuery`, which **again** uses `myProfile` GraphQL query to fetch the profile data. It should be enough to use the `myProfile` GraphQL query only once.

Thx, this should be changed too! At the moment the ProfileContext cannot handle this scenario. It only loads, keeps and mutates complete profile data. But it cannot handle the scenario, where it profile does not exist. This raise an error. And it cannot reload the profile after is has been created. 

So, I'll make a ticket for this and change it later.

NikoHelle commented:
FYI: I added a new commit regarding double loading of the api tokens.

akikoskinen reviewed (approved):
:+1:
